sig-scalability/blogs/k8s-services-scalability-issues.md

```diff
@@ -0,0 +1,130 @@
+
+#### Possible Solution(s)
+
+- Redesign endpoints API. Some properties we may like to have for the new API:
+  - Lives in an API group
+  - Not coupled to service resource lifecycle
+  - Can track both pod and non-pod endpoints
+  - Sharded into multiple etcd key/value pairs? (potentially in its own etcd)
+  - Stable sharding & ordering, with a ready property instead of separate lists
+  - Port ranges and other features we’ve put off for a long time
+
+
+## Endpoints read traffic is (super-)quadratic in the no. of endpoints
+
+#### Issue
+
+This is the equivalent issue for 'reads' as the above one which was for 'writes'. An endpoints read would return all the O(N) endpoints present in the object - not allowing us to efficiently retrieve only a single endpoint we may be interested in. This becomes an even bigger problem with watches on endpoints (e.g ones from kube-proxies) as the quadratic updates get multiplied further with a factor of #watches (which is at least O(#nodes)).
+
+#### Relevant issues/links
+
+- [#24552](https://github.com/kubernetes/kubernetes/issues/24552)
+- [#8190](https://github.com/kubernetes/kubernetes/issues/8190)
+
+#### Possible Solutions(s)
+
+- Redesign endpoints API (as suggested also above)
+
+
+## Define/measure performance SLIs for endpoints propagation
+
+#### Issue
+
+We need to measure the e2e endpoints propagation latency, i.e the time from when a 
+service is created until its endpoints are populated, given that all of the associated pods 
+are already running and ready. We also need to come up with reasonable SLOs for the 
+same and verify that they’re satisfied at scale.
+
+#### Relevant issues/links
+
+- [#10436](https://github.com/kubernetes/kubernetes/issues/10436)
+- This [doc on networking SLIs](https://docs.google.com/document/d/1kudI8uILO13ySkpE5HgNHvIGn-yXVQatpyM3YTVVUis) proposes one way of formulating the SLI
+
+#### Possible Solution(s)
+
+- Measure the SLI defined in the above doc and come up with a reasonable SLO
+
+
+## Master election shouldn’t consume unnecessary watch bandwidth
+
+#### Issue
+
+Currently we store the leader lock for master components like the scheduler in the corresponding endpoints object. This has the undesirable side effect of sending a notification down the kube-proxy <-> master (and kube-dns <-> master) watch every second, which is not actually required.
+
+#### Relevant issues/links
+
+- [#34627](https://github.com/kubernetes/kubernetes/issues/34627)
+
+#### Possible Solutions(s)
+
+- Avoid overloading endpoints as a lock API - we can use configmaps instead, which is a supported option already in our client-go leader election library. One tricky problem here is performing upgrades of HA clusters that are using the old endpoints-based locking approach to this configmaps-based approach without breaking them.
+- Filter those watches to not get such updates. This is more of a hack than a proper solution, so we choose not to do it.
```

Comment:
We shouldn't even mention that - we need to solve it properly.

Reply:
Removed it.